Add nxml-uxml

dpk opened the pull request:
### Brief summary of what the package does

[MicroXML](https://dvcs.w3.org/hg/microxml/raw-file/tip/spec/microxml.html) is a delightful tiny subset of XML, removing everything that
makes XML a nightmare to deal with in practice. `nxml-uxml-mode` is a
minor mode for Emacs that slightly modifies the XML parser of
`nxml-mode` to forbid most (though, at present, not quite all)
constructs which are allowed in XML 1.0 but disallowed in MicroXML.


### Direct link to the package repository

https://gitlab.com/dpk/nxml-uxml

### Your association with the package

Author and maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm with `x`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
  - It complains about the function names starting with `uxmltok-` because they don’t match the package name. Since these actually patch the `xmltok` library, I think this is probably a reasonable exception to this rule, but I’ll change it if the MELPA maintainers disagree.
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this!  MicroXML looks like something I should check out.  Quick first pass below --

`nxml-uxml.el` with byte-compile using Emacs 27.1:
```
nxml-uxml.el:118:20:Warning: reference to free variable `xmltok-type'
nxml-uxml.el:119:20:Warning: reference to free variable `xmltok-name-colon'
nxml-uxml.el:133:38:Warning: reference to free variable `xmltok-start'
nxml-uxml.el:121:51:Warning: reference to free variable `xmltok-name-end'
nxml-uxml.el:122:38:Warning: reference to free variable `xmltok-attributes'
nxml-uxml.el:122:56:Warning: reference to free variable
    `xmltok-namespace-attributes'
In end of data:
nxml-uxml.el:159:1:Warning: the following functions are not known to be
    defined: rng-after-change-function, xmltok-add-error
```
I think some of these need a `(require 'xmltok)` and maybe `(require 'rng-nxml)` although I'm not familiar with these two buildin dependencies.

`nxml-uxml.el` with checkdoc 0.6.2:
```
nxml-uxml.el:52: Messages should start with a capital letter
```
Extremely minor - some people solve this by writing ``` `nxml-uxml-mode' can only be ... ``` for example.

`nxml-uxml.el` with package-lint 20211122.1152:
```
5 issues found:
1:0: warning: "Version:" or "Package-Version:" header is missing. MELPA will handle this, but other archives will not.
65:0: error: "uxmltok-disallow-in-prolog" doesn't start with package's prefix "nxml-uxml".
91:0: error: "uxmltok-disallow-gts" doesn't start with package's prefix "nxml-uxml".
100:0: error: "uxmltok-disallow-newlines" doesn't start with package's prefix "nxml-uxml".
111:0: error: "uxmltok-disallow-in-content" doesn't start with package's prefix "nxml-uxml".
```

`nxml-uxml.el` with [melpazoid](https://github.com/riscy/melpazoid):
- nxml-uxml.el#L49: Lighter should start with a space
- nxml-uxml.el#L89: Loading a package should rarely add advice -- should this sort of thing be gated behind a mode toggle?  Similarly for line 153

Package and license:
- Add a LICENSE file to the repository

dpk commented:
I have fixed the free variable and undefined function warnings and the problem with the error message and lighter as suggested, and added a licence file.

> ```
> 1:0: warning: "Version:" or "Package-Version:" header is missing. MELPA will handle this, but other archives will not.
> ```

Am I to take it that you mentioning this means I’ve misunderstood what ‘MELPA will handle this’ implies? Does MELPA (by policy) require a version even if it can (technically) handle it being missing?

> ```
> 65:0: error: "uxmltok-disallow-in-prolog" doesn't start with package's prefix "nxml-uxml".
> 91:0: error: "uxmltok-disallow-gts" doesn't start with package's prefix "nxml-uxml".
> 100:0: error: "uxmltok-disallow-newlines" doesn't start with package's prefix "nxml-uxml".
> 111:0: error: "uxmltok-disallow-in-content" doesn't start with package's prefix "nxml-uxml".
> ```

Are you then rejecting my argument for these being a reasonable exception to this rule because they advise the `xmltok` package, not `nxml`?

> nxml-uxml.el#L89: Loading a package should rarely add advice -- should this sort of thing be gated behind a mode toggle? Similarly for line 153

The actual advice is written to have no effect unless the `nxml-uxml-mode` minor mode is active in the current buffer. Advice can’t be buffer-local in any other way, as far as I can tell. Or should adding the advice be delayed until `nxml-uxml-mode` is actually activated? It doesn’t make much functional difference either way. (Note that such ‘buffer-local’ advice can’t be removed when the minor mode is deactivated, because there might be other buffers using it by then.)

riscy commented:
> Am I to take it that you mentioning this means I’ve misunderstood what ‘MELPA will handle this’ implies? Does MELPA (by policy) require a version even if it can (technically) handle it being missing?

In this case we generally recommend you put a `Version:` (MELPA auto-generates `Package-Version`).  It's true the MELPA code _will_ handle this.  I think even purcell -- who maintains package-lint -- just puts `Version: 0.0` or similar in his packages and calls it a day.  Sort of an explicit opt-out.

> Are you then rejecting my argument for these being a reasonable exception to this rule because they advise the xmltok package, not nxml?

This is my fault for not reading the PR summary carefully.  Lining the function prefixes with the package name is a simple and fair (if imperfect) way to prevent namespace collisions across MELPA (and hopefully elsewhere).  In very rare cases I would make exceptions to this rule (e.g. backward compatibility with a package predating this rule) -- and `package-lint` has some of these exceptions built into it -- but that's the reason why package-lint is so strict about it.  We generally prefer packages to try to adhere when they can.

> should adding the advice be delayed until nxml-uxml-mode is actually activated? It doesn’t make much functional difference either way.

Yeah, that's what I'd recommend.  It's a blunt suggestion, but it makes things easier en masse. :)

dpk commented:
Okay, I’ve added a fake `Version:` as suggested, renamed the offending functions, and moved adding advice into the mode initialization function.

riscy commented:
Thanks for accommodating -- everything looks good to go.  We also strongly suggest adding a LICENSE file to the repository (one that GitLab is able to detect) just to cover your bases, but I'll spare you the back and forth over that.

Welcome to MELPA! :)
